### PR TITLE
feat(checks): add secrets leak check in Dockerfile

### PR DESCRIPTION
**Pull request by nikpivkin**
Close https://github.com/aquasecurity/trivy/issues/7639

The check is triggered in the following cases:
1. Using a secret environment variable in the argument (ARG) or environment variable (ENV).

Dockerfile:
```Dockerfile
FROM alpine
ARG GITHUB_TOKEN
```
Output:
```bash
CRITICAL: Possible exposure of secret env "GITHUB_TOKEN" in ARG
═════════════════════════════════════════════════════════════════════════════════════════════════════════════════════
Passing secrets via `build-args` or envs or copying secret files can leak them out

See https://avd.aquasec.com/misconfig/ds031
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 Dockerfile:4
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
   4 [ ARG GITHUB_TOKEN
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
```
2. Copying a secret file, such as an AWS credential file, to an image.

Dockerfile:
```Dockerfile
FROM alpine
ENV AWS_CONFIG_FILE=/.aws/config
COPY ./config $AWS_CONFIG_FILE
```

Output:
```bash
CRITICAL: Possible exposure of secret file "AWS_CONFIG_FILE" in COPY
═════════════════════════════════════════════════════════════════════════════════════════════════════════════════════
Passing secrets via `build-args` or envs or copying secret files can leak them out

See https://avd.aquasec.com/misconfig/ds031
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 Dockerfile.dev1:3
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
   3 [ COPY ./config $AWS_CONFIG_FILE
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
```

Or

```Dockerfile
FROM alpine
COPY . .
ENV GOOGLE_APPLICATION_CREDENTIALS="./news-extraction.json"
```

3. Setting credentials via the setup command.
Dockerfile:
```Dockerfile
FROM amazon/aws-cli:latest

RUN aws configure set aws_access_key_id test-id && \
    aws configure set aws_secret_access_key test-key
```

Output:
```bash
CRITICAL: Possible exposure of secret in RUN
═══════════════════════════════════════════════════════════════════════════════════════════════════════════
Passing secrets via `build-args` or envs or copying secret files can leak them out

See https://avd.aquasec.com/misconfig/ds031
───────────────────────────────────────────────────────────────────────────────────────────────────────────
 Dockerfile.dev5:3-4
───────────────────────────────────────────────────────────────────────────────────────────────────────────
   3 ┌ RUN aws configure set aws_access_key_id test-id && \
   4 └     aws configure set aws_secret_access_key test-key
───────────────────────────────────────────────────────────────────────────────────────────────────────────
```

This ignores the instructions that the secret mount uses:
```Dockerfile
# syntax=docker/dockerfile:1

FROM python:3.9-slim

RUN apt-get update && apt-get install -y curl unzip \
    && curl "https://awscli.amazonaws.com/awscli-exe-linux-x86_64.zip" -o "awscliv2.zip" \
    && unzip awscliv2.zip \
    && ./aws/install

RUN --mount=type=secret,id=aws-key-id,env=AWS_ACCESS_KEY_ID \
    --mount=type=secret,id=aws-secret-key,env=AWS_SECRET_ACCESS_KEY \
    aws configure set aws_access_key_id $(cat /run/secrets/aws-key-id) && \
    aws configure set aws_secret_access_key $(cat /run/secrets/aws-secret-key)
```

### CODE REVIEW

**Comment by nikpivkin** [comment]
@simar7 Do we need to create two separate checks for secret checking and password detection?

File: checks/docker/leaked_secrets.rego [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+# schemas:
+# - input: schema["dockerfile"]
+# related_resources:
+# - https://docs.docker.com/build/building/secrets/
+# custom:
+#   id: DS031
+#   avd_id: AVD-DS-0031
+#   severity: CRITICAL
+#   short_code: do-not-pass-secrets
+#   recommended_action: Use secret mount if secrets are needed during image build. Use volume mount if secret files are needed during container runtime.
```
Comment (simar7) [inline checks/docker/leaked_secrets.rego:13]:
```suggestion
#   recommended_action: Use secret mount if secrets are needed during image build. Use volume mount if secured files are needed during container runtime.
```

File: checks/docker/leaked_secrets.rego [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+#   input:
+#     selector:
+#     - type: dockerfile
+package builtin.dockerfile.DS031
+
+import rego.v1
+
+import data.lib.docker
+
+final_stage := last(input.Stages)
```
Comment (simar7) [inline checks/docker/leaked_secrets.rego:23]: Can this list be assumed to be always sorted?

File: cmd/id/main.go [hunk 3]
```diff
@@ -8,12 +8,15 @@ import (
...
 	"github.com/aquasecurity/trivy/pkg/iac/framework"
+	"github.com/aquasecurity/trivy/pkg/iac/rego"
 	_ "github.com/aquasecurity/trivy/pkg/iac/rego"
 	"github.com/aquasecurity/trivy/pkg/iac/rules"
 )
 
 func main() {
 
+	rules.Reset()
+	rego.LoadAndRegister()
```
Comment (simar7) [inline cmd/id/main.go:19]: Curious why this change?

File: checks/docker/leaked_secrets.rego [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+
+remove_trailing_slash(path) := cut_suffix(path, "/")
+
+cut_prefix(s, prefix) := substring(s, 1, -1) if {
+	startswith(s, prefix)
+} else := s
+
+cut_suffix(s, suffix) := substring(s, 0, count(s) - 1) if {
+	endswith(s, suffix)
+} else := s
```
Comment (simar7) [inline checks/docker/leaked_secrets.rego:162]: Should we move some of them to the library?

File: checks/docker/leaked_secrets.rego [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+	)
+}
+
+check_args := true
+
+# TODO: Should arguments be checked?
+is_arg_or_env(cmd) if {
+	check_args
+	cmd == "arg"
+}
```
Comment (simar7) [inline checks/docker/leaked_secrets.rego:70]: what's a case that you are thinking to check the args for?

File: cmd/id/main.go (hunk 3 above)
Comment (nikpivkin) [reply cmd/id/main.go:19]: This only loads Rego checks.

File: checks/docker/leaked_secrets.rego (hunk 1 above)
Comment (nikpivkin) [reply checks/docker/leaked_secrets.rego:13]: why “secured”?

File: checks/docker/leaked_secrets.rego (hunk 4 above)
Comment (nikpivkin) [reply checks/docker/leaked_secrets.rego]: Done https://github.com/aquasecurity/trivy-checks/pull/265/commits/f97bc565b22d4af4c5f27fe79513b284fbd784d9

File: checks/docker/leaked_secrets.rego (hunk 5 above)
Comment (nikpivkin) [reply checks/docker/leaked_secrets.rego]: Made it enabled by default https://github.com/aquasecurity/trivy-checks/pull/265/commits/9fc294b5d2bbd232097b12c9d92e3bbf23d2eeef

File: checks/docker/leaked_secrets.rego (hunk 2 above)
Comment (nikpivkin) [reply checks/docker/leaked_secrets.rego:24]: Dockerfile is parsed from top to bottom, so the order of stages is preserved. The order of arrays is also preserved when marshaling json (export to Rego).

**Comment by nikpivkin** [comment]
@simar7 I added the ability to add custom environment variables to easily extend the checkout without having to open PR.

```bash
❯ cat my-envs.yaml
ds031:
  included_envs: ["MY_SECRET"]

❯ cat Dockerfile.custom
from alpine

arg MY_SECRET

❯ trivy conf -d Dockerfile.custom --config-data my-envs.yaml
CRITICAL: Possible exposure of secret env "MY_SECRET" in ARG
═══════════════════════════════════════════════════════════════════════════════════════════════════════════
Passing secrets via `build-args` or envs or copying secret files can leak them out

See https://avd.aquasec.com/misconfig/ds031
───────────────────────────────────────────────────────────────────────────────────────────────────────────
 Dockerfile.custom:3
───────────────────────────────────────────────────────────────────────────────────────────────────────────
   3 [ arg MY_SECRET
───────────────────────────────────────────────────────────────────────────────────────────────────────────
```

This can also be implemented for [secret tokens](https://github.com/aquasecurity/trivy-checks/pull/265/files#diff-507f7dc2d5dcaf9606a337ba939a05532e5402f5f3bf00c22bbfe627812437c2R221-R225) if there is a need.

**Review by simar7 — APPROVED** [review]

**Comment by irbull** [comment]
We just got bit by this today. It's not clear why this just happened today as this was merged 2 weeks ago.

However, the URL it suggests to visit for more information, returns a 404: https://avd.aquasec.com/misconfig/ds031.

Is that a separate issue?
